File: clients/python/src/hiero_proxy/resources/accounts.py
"""Accounts resource."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from hiero_proxy.resources.validation import require_id

if TYPE_CHECKING:
    from hiero_proxy.client import AsyncHieroProxyClient, HieroProxyClient


class AccountsResource:
    """Account operations — create, query, update, delete, transfer."""

    def __init__(self, client: HieroProxyClient | AsyncHieroProxyClient) -> None:
        self._client = client

    def create(self, initial_balance_in_hbar: int | None = None) -> Any:
        body = {"initialBalanceInHbar": initial_balance_in_hbar} if initial_balance_in_hbar is not None else None
        return self._client.post("/api/v1/accounts", json=body)

    def get_balance(self, account_id: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.get(f"/api/v1/accounts/{aid}/balance")

    def get_operator_balance(self) -> Any:
        return self._client.get("/api/v1/accounts/operator/balance")

    def get_info(self, account_id: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.get(f"/api/v1/accounts/{aid}")

    def update_key(self, account_id: str, current_private_key: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.put(f"/api/v1/accounts/{aid}/key", json={"currentPrivateKey": current_private_key})

    def update_memo(self, account_id: str, private_key: str, memo: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.put(f"/api/v1/accounts/{aid}/memo", json={"privateKey": private_key, "memo": memo})

    def update(self, account_id: str, current_private_key: str, memo: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.put(f"/api/v1/accounts/{aid}", json={"currentPrivateKey": current_private_key, "memo": memo})

    def delete(self, account_id: str, private_key: str) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.delete(f"/api/v1/accounts/{aid}", json={"privateKey": private_key})

    def delete_to_recipient(self, account_id: str, recipient_account_id: str, private_key: str) -> Any:
        aid = require_id(account_id, "account_id")
        rid = require_id(recipient_account_id, "recipient_account_id")
        return self._client.delete(f"/api/v1/accounts/{aid}/to/{rid}", json={"privateKey": private_key})

    def transfer_from_operator(self, account_id: str, amount: int) -> Any:
        aid = require_id(account_id, "account_id")
        return self._client.post(f"/api/v1/accounts/{aid}/transfer/operator", json={"amount": amount})

    def transfer(self, from_account_id: str, to_account_id: str, amount: int, private_key: str) -> Any:
        fid = require_id(from_account_id, "from_account_id")
        return self._client.post(
            f"/api/v1/accounts/{fid}/transfer",
            json={"toAccountId": to_account_id, "amount": amount, "privateKey": private_key},
        )

File: clients/javascript/src/resources/accounts.ts
import type { HieroProxyClient } from "../client";
import type {
  AccountResponse,
  BalanceResponse,
  AccountInfoResponse,
  SuccessResponse,
  CreateAccountRequest,
  UpdateKeyRequest,
  UpdateMemoRequest,
  UpdateAccountRequest,
  DeleteAccountRequest,
  DeleteAccountToRecipientRequest,
  OperatorTransferRequest,
  TransferRequest,
} from "./types";

/** Validates that an account ID is a non-empty string matching the Hiero format. */
function requireAccountId(accountId: string, paramName = "accountId"): string {
  if (!accountId || typeof accountId !== "string" || !accountId.trim()) {
    throw new Error(`${paramName} is required and must be a non-empty string`);
  }
  return encodeURIComponent(accountId.trim());
}

/**
 * Accounts resource — create, query, update, delete accounts and transfer HBAR.
 */
export class AccountsResource {
  constructor(private readonly client: HieroProxyClient) {}

  /**
   * Create a new account on the Hiero network.
   * @param initialBalanceInHbar Optional initial HBAR balance (default: 0).
   */
  async create(initialBalanceInHbar?: number): Promise<AccountResponse> {
    const body: CreateAccountRequest | undefined =
      initialBalanceInHbar !== undefined ? { initialBalanceInHbar } : undefined;
    return this.client.post("/api/v1/accounts", body);
  }

  /**
   * Get the HBAR balance of a specific account.
   */
  async getBalance(accountId: string): Promise<BalanceResponse> {
    const id = requireAccountId(accountId);
    return this.client.get(`/api/v1/accounts/${id}/balance`);
  }

  /**
   * Get the operator account's HBAR balance.
   */
  async getOperatorBalance(): Promise<BalanceResponse> {
    return this.client.get("/api/v1/accounts/operator/balance");
  }

  /**
   * Get detailed account info from the mirror node.
   */
  async getInfo(accountId: string): Promise<AccountInfoResponse> {
    const id = requireAccountId(accountId);
    return this.client.get(`/api/v1/accounts/${id}/info`);
  }

  /**
   * Rotate the key pair of an account. Returns the new key pair.
   * @param accountId The account to update.
   * @param currentPrivateKey The current private key to authorize the rotation.
   */
  async updateKey(accountId: string, currentPrivateKey: string): Promise<AccountResponse> {
    const id = requireAccountId(accountId);
    const body: UpdateKeyRequest = { currentPrivateKey };
    return this.client.put(`/api/v1/accounts/${id}/key`, body);
  }

  /**
   * Update the memo field of an account.
   * @param accountId The account to update.
   * @param privateKey The account's private key.
   * @param memo The new memo value.
   */
  async updateMemo(accountId: string, privateKey: string, memo: string): Promise<SuccessResponse> {
    const id = requireAccountId(accountId);
    const body: UpdateMemoRequest = { privateKey, memo };
    return this.client.put(`/api/v1/accounts/${id}/memo`, body);
  }

  /**
   * Update both the key and memo of an account atomically.
   */
  async update(accountId: string, currentPrivateKey: string, memo: string): Promise<AccountResponse> {
    const id = requireAccountId(accountId);
    const body: UpdateAccountRequest = { currentPrivateKey, memo };
    return this.client.put(`/api/v1/accounts/${id}`, body);
  }

  /**
   * Delete an account. Remaining balance goes to the operator.
   */
  async delete(accountId: string, privateKey: string): Promise<SuccessResponse> {
    const id = requireAccountId(accountId);
    const body: DeleteAccountRequest = { privateKey };
    return this.client.delete(`/api/v1/accounts/${id}`, body);
  }

  /**
   * Delete an account. Remaining balance goes to a specified recipient.
   */
  async deleteToRecipient(
    accountId: string,
    recipientAccountId: string,
    accountPrivateKey: string,
    recipientPrivateKey: string
  ): Promise<SuccessResponse> {
    const id = requireAccountId(accountId);
    const recipientId = requireAccountId(recipientAccountId, "recipientAccountId");
    const body: DeleteAccountToRecipientRequest = { accountPrivateKey, recipientPrivateKey };
    return this.client.delete(`/api/v1/accounts/${id}/to/${recipientId}`, body);
  }

  /**
   * Transfer HBAR from the operator to a target account.
   */
  async transferFromOperator(toAccountId: string, amountInHbar: number): Promise<SuccessResponse> {
    const body: OperatorTransferRequest = { toAccountId, amountInHbar };
    return this.client.post("/api/v1/accounts/transfer", body);
  }

  /**
   * Transfer HBAR between user accounts.
   */
  async transfer(
    fromAccountId: string,
    fromAccountPrivateKey: string,
    toAccountId: string,
    amountInHbar: number
  ): Promise<SuccessResponse> {
    const id = requireAccountId(fromAccountId, "fromAccountId");
    const body: TransferRequest = { fromAccountPrivateKey, toAccountId, amountInHbar };
    return this.client.post(`/api/v1/accounts/${id}/transfer`, body);
  }
}
